File: matrix_multiply.cpp
#include <cstdlib>
#include <ctime>
#include <iostream>
#include <numeric>


const int SIZE = 1000;  // size of the square matrices


// generate and fill a matrix with random integers (0-9)
void generateMatrix(int matrix[SIZE][SIZE]) {
    static bool seeded = false;
    if (!seeded) {
        std::srand(std::time(0));
        seeded = true;
    }
    for (int i = 0; i < SIZE; ++i) {
        for (int j = 0; j < SIZE; ++j) {
            matrix[i][j] = std::rand() % 10;
        }
    }
}


// multiply matrixA and matrixB and store the result in resultMatrix
void multiplyMatrices(
    const int matrixA[SIZE][SIZE],
    const int matrixB[SIZE][SIZE],
    volatile int resultMatrix[SIZE][SIZE]
) {
    for (int i = 0; i < SIZE; ++i) {
        for (int j = 0; j < SIZE; ++j) {
            resultMatrix[i][j] = 0;
            for (int k = 0; k < SIZE; ++k) {
                resultMatrix[i][j] += matrixA[i][k] * matrixB[k][j];
            }
        }
    }
}


int main() {
    // move these 16 MB variables off the stack and into static memory
    // to prevent a stack overflow segmentation fault
    static int A[SIZE][SIZE];
    static int B[SIZE][SIZE];

    // keep `volatile` to prevent compiler from optimizing the math
    static volatile int C[SIZE][SIZE];

    // generate random values for A and B
    generateMatrix(A);
    generateMatrix(B);

    // perform the matrix multiplication
    multiplyMatrices(A, B, C);

    long long checksum = 0;
    for (int i = 0; i < SIZE; ++i) {
        for (int j = 0; j < SIZE; ++j) {
            checksum += C[i][j];
        }
    }
    std::cerr << "checksum = " << checksum << std::endl;

    return 0;
}

File: matrix_multiply.py
import cProfile
import random
import sys
import time


SIZE = 1000  # size of the square matrices


def generate_matrix():
    """
    Generate a SIZE x SIZE matrix (list of lists) with random integers
    between 0 and 9.
    """
    matrix = [[0] * SIZE for _ in range(SIZE)]
    for i in range(SIZE):
        for j in range(SIZE):
            matrix[i][j] = random.randint(0, 9)
    return matrix


def multiply_matrices(matrix_a, matrix_b):
    """
    Perform standard (cubic-time) matrix multiplication (A * B = C).
    """
    result_matrix = [[0] * SIZE for _ in range(SIZE)]

    for i in range(SIZE):           # rows of the result matrix (i)
        for j in range(SIZE):       # columns of the result matrix (j)
            for k in range(SIZE):   # columns of A and rows of B (k)
                result_matrix[i][j] += matrix_a[i][k] * matrix_b[k][j]
                # C[i][j] += A[i][k] * B[k][j], this line involves:
                #   three Python integer lookups
                #   one multiplication
                #   one addition
                #   one assignment

    return result_matrix


def main():
    A = generate_matrix()
    B = generate_matrix()
    C = multiply_matrices(A, B)

    checksum = 0
    for i in range(SIZE):
        checksum += sum(C[i])
    print(f"{checksum = }")


########################################
# test runtime
########################################


random.seed(time.time())


# without profiling
start = time.time()
main()
end = time.time()
print(f"total time = {end - start:.5f} seconds")


# with profiling
# profiler = cProfile.Profile()
# profiler.enable()
# main()
# profiler.disable()
# profiler.print_stats()
